# Exploratory Data Analysis on MovieLens Dataset

> Kunal PATIL (AIS S20)

In this notebook we are going to do exploratory Data analysis on Movielens dataset.
We will be performing below explorations:
1. Explaining the data
2. Cleaning the data 
(Identifying duplicates, Eliminating duplicates if there are any and Adapting these changes in ratings dataset to avoid wrong interpretation of data)

In [1]:
!python --version

Python 3.8.7


### Arborescence

In [3]:
!tree ../

Folder PATH listing for volume PROJECT
Volume serial number is E8EA-F1D3
D:\SUBJECTS\RECSYS\PROJECT\
Invalid path - \SUBJECTS\RECSYS\PROJECT\
No subfolders exist 



### Import Libraries and Declare variables

In [3]:
from pathlib import Path
import pandas as pd

In [4]:
DATA_DIR = Path('../data/movielens/ml-latest')

RATINGS_FILEPATH = DATA_DIR / 'ratings.csv'
MOVIES_FILPATH = DATA_DIR / 'movies.csv'

# Ratings

#### Performing analysis on ratings dataset.

Read data using pandas and save it in variable

In [5]:
ratings_raw = pd.read_csv(RATINGS_FILEPATH)
ratings_raw.sample(5)

userId  movieId  rating   timestamp
8573461    88366    36529     3.0  1326719775
10539337  108312    33493     1.5  1221972955
14844458  151859     1527     4.0   966129091
15501777  158274    49530     3.5  1461063353
26024629  265806    26776     5.0  1170141136

> There are four columns UserId, movieId, rating and timestamp, showing only first 5 rows.

In [6]:
# Make a duplicate of original dataset for further operations
ratings = ratings_raw.copy()

In [7]:
# Check the information about the dataset
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27753444 entries, 0 to 27753443
Data columns (total 4 columns):
 #   Column     Dtype  
---  ------     -----  
 0   userId     int64  
 1   movieId    int64  
 2   rating     float64
 3   timestamp  int64  
dtypes: float64(1), int64(3)
memory usage: 847.0 MB


Check if there is any null data present in any column

In [8]:
ratings.isnull().sum()

userId       0
movieId      0
rating       0
timestamp    0
dtype: int64

> There is no null values present in our dataset, So we have our all data, but what if we have duplicate entries?

Check whether is there any duplicate rows present and if yes what's the count.

In [9]:
ratings.duplicated(keep=False).sum()

0

> There seems no duplication in row values, means all movieIds are distinct, but again what if there are duplication in other columns like ratings, userId and timestamp?

Check the duplicates per column if any

In [11]:
ratings.duplicated(subset=['userId', 'movieId', 'rating', 'timestamp']).sum()

0

> There are no duplicates found.

We will check movie dataset and we will later come back to ratings dataset if necessary.

# Movies

#### Performing analysis on movies dataset.

Read data using pandas and save it in variable

In [12]:
movies_raw = pd.read_csv(MOVIES_FILPATH)
movies_raw.sample(5)

movieId                          title                genres
20818   101162  Last Call at the Oasis (2011)           Documentary
19491    96110           Campaign, The (2012)                Comedy
55266   186953        Hearts Beat Loud (2018)                 Drama
34169   139397    Peace After Marriage (2013)  Comedy|Drama|Romance
49462   174171  The Kingdom of Fairies (1903)     Adventure|Fantasy

> There are three columns movieId, Title and Genre, showing only first 5 rows.

In [13]:
# Make a duplicate of original dataset for further operations
movies = movies_raw.copy()

In [14]:
# Check the information about the dataset
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 58098 entries, 0 to 58097
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   movieId  58098 non-null  int64 
 1   title    58098 non-null  object
 2   genres   58098 non-null  object
dtypes: int64(1), object(2)
memory usage: 1.3+ MB


Check if there is any null data present in any column

In [15]:
 movies.isnull().sum()

movieId    0
title      0
genres     0
dtype: int64

> There is no null values present in our dataset, So we have our all data, but what if we have duplicate entries?

Check whether is there any duplicate rows present and if yes what's the count.

In [16]:
movies.duplicated().sum()

0

> There seems no duplication in row values, means all movieIds are distinct, but again what if there are duplication in other columns like with movie title or genres.

Check the duplicates per column if any

In [17]:
movies.duplicated(subset=['title', 'genres']).sum()

14

> Yes! There are some duplicates present, let's check further in details

Check the duplicates associated with columns, title and genres

In [18]:
duplicates = movies[movies.duplicated(subset=['title', 'genres'], keep=False)]
duplicates

movieId                        title                   genres
2560      2644               Dracula (1931)                   Horror
11173    46865            Little Man (2006)                   Comedy
11412    48682               Offside (2006)             Comedy|Drama
13504    66511        Berlin Calling (2008)             Comedy|Drama
14038    70155  Seven Years Bad Luck (1921)                   Comedy
15637    79254                  Home (2008)                    Drama
15902    80330               Offside (2006)             Comedy|Drama
18867    93279               Darling (2007)                    Drama
19914    97773             Girl, The (2012)                    Drama
20835   101212             Girl, The (2012)                    Drama
21655   104035               Beneath (2013)                   Horror
21691   104155         Clear History (2013)                   Comedy
23834   111519        Johnny Express (2014)  Animation|Comedy|Sci-Fi
25046   115777               Beneath (2013)                   Horror
27572   122940         Clear History (2013)                   Comedy
29852   128991        Johnny Express (2014)  Animation|Comedy|Sci-Fi
30226   130062               Darling (2007)                    Drama
33001   136564               Macbeth (2015)                    Drama
36172   143978                  Home (2008)                    Drama
38804   150310               Macbeth (2015)                    Drama
44387   163246  Seven Years Bad Luck (1921)                   Comedy
46904   168774                Detour (2017)                 Thriller
48620   172427            Little Man (2006)                   Comedy
51339   178401                 Lucky (2017)                    Drama
52192   180205               Dracula (1931)                   Horror
52644   181329                 Lucky (2017)                    Drama
52795   181655                Detour (2017)                 Thriller
57269   191775        Berlin Calling (2008)             Comedy|Drama

> We can see there are 14 pair of entries having same movie title and genre but different movieIds.

Check values of duplicated movie titles column

In [19]:
duplicates.title.values

array(['Dracula (1931)', 'Little Man (2006)', 'Offside (2006)',
       'Berlin Calling (2008)', 'Seven Years Bad Luck (1921)',
       'Home (2008)', 'Offside (2006)', 'Darling (2007)',
       'Girl, The (2012)', 'Girl, The (2012)', 'Beneath (2013)',
       'Clear History (2013)', 'Johnny Express (2014)', 'Beneath (2013)',
       'Clear History (2013)', 'Johnny Express (2014)', 'Darling (2007)',
       'Macbeth (2015)', 'Home (2008)', 'Macbeth (2015)',
       'Seven Years Bad Luck (1921)', 'Detour (2017)',
       'Little Man (2006)', 'Lucky (2017)', 'Dracula (1931)',
       'Lucky (2017)', 'Detour (2017)', 'Berlin Calling (2008)'],
      dtype=object)

Display duplicate values among the unique values from the titles and pair them per duplication

In [20]:
for movie_title in duplicates.title.unique():
    display(movies[movies.title == movie_title])

movieId           title  genres
2560      2644  Dracula (1931)  Horror
52192   180205  Dracula (1931)  Horror

movieId              title  genres
11173    46865  Little Man (2006)  Comedy
48620   172427  Little Man (2006)  Comedy

movieId           title        genres
11412    48682  Offside (2006)  Comedy|Drama
15902    80330  Offside (2006)  Comedy|Drama

movieId                  title        genres
13504    66511  Berlin Calling (2008)  Comedy|Drama
57269   191775  Berlin Calling (2008)  Comedy|Drama

movieId                        title  genres
14038    70155  Seven Years Bad Luck (1921)  Comedy
44387   163246  Seven Years Bad Luck (1921)  Comedy

movieId        title genres
15637    79254  Home (2008)  Drama
36172   143978  Home (2008)  Drama

movieId           title genres
18867    93279  Darling (2007)  Drama
30226   130062  Darling (2007)  Drama

movieId             title genres
19914    97773  Girl, The (2012)  Drama
20835   101212  Girl, The (2012)  Drama

movieId           title  genres
21655   104035  Beneath (2013)  Horror
25046   115777  Beneath (2013)  Horror

movieId                 title  genres
21691   104155  Clear History (2013)  Comedy
27572   122940  Clear History (2013)  Comedy

movieId                  title                   genres
23834   111519  Johnny Express (2014)  Animation|Comedy|Sci-Fi
29852   128991  Johnny Express (2014)  Animation|Comedy|Sci-Fi

movieId           title genres
33001   136564  Macbeth (2015)  Drama
38804   150310  Macbeth (2015)  Drama

movieId          title    genres
46904   168774  Detour (2017)  Thriller
52795   181655  Detour (2017)  Thriller

movieId         title genres
51339   178401  Lucky (2017)  Drama
52644   181329  Lucky (2017)  Drama

> When listed movies according to titles we can see each movie have same title and genres but different movieId

So this can be observed that if there are duplicated movies present in movies dataset but with different movieIds, so what about their entries in ratings dataset. In movies dataset, same movie is associated with different movieIds so if we dont process ratings along with movie dataset, it will give the wrong input for processing our system.

As we saved duplicates, select movieIds and compare with ratings dataset from which we have to process the duplicated entries.

In [21]:
nb_ratings_to_drop = 0
for movie_id in duplicates.movieId.values:
    nb_ratings_to_drop += (ratings.movieId == movie_id).sum()
nb_ratings_to_drop

3762

> So we have to process 3762 entries present in ratings dataset to avoid duplication

Check the total number of movie titles are duplicated

In [22]:
movies.title.duplicated().sum()

78

In [23]:
movies[movies.title.duplicated(keep=False)]

movieId                                      title  \
582        588                             Aladdin (1992)   
823        838                                Emma (1996)   
1716      1788                       Men with Guns (1997)   
2560      2644                             Dracula (1931)   
2766      2851                            Saturn 3 (1980)   
...        ...                                        ...   
57269   191775                      Berlin Calling (2008)   
57305   191867                  Let There Be Light (2017)   
57361   192003  Journey to the Center of the Earth (2008)   
57463   192243                             Contact (1992)   
57953   193559                               Cargo (2017)   

                                            genres  
582    Adventure|Animation|Children|Comedy|Musical  
823                           Comedy|Drama|Romance  
1716                                  Action|Drama  
2560                                        Horror  
2766                     Adventure|Sci-Fi|Thriller  
...                                            ...  
57269                                 Comedy|Drama  
57305                                  Documentary  
57361              Action|Adventure|Fantasy|Sci-Fi  
57463                Drama|Horror|Mystery|Thriller  
57953                                        Drama  

[156 rows x 3 columns]

Before dropping any values first check whether we are following correct approach to get the duplicates of the movies. Testing our logic on single entry before building a function to process dataset.

In [24]:
movies[movies.title == 'Berlin Calling (2008)']

movieId                  title        genres
13504    66511  Berlin Calling (2008)  Comedy|Drama
57269   191775  Berlin Calling (2008)  Comedy|Drama

Check in ratings the particular movieId (eg. 66511) is repeated how many times.

In [25]:
ratings[ratings.movieId == 66511]

userId  movieId  rating   timestamp
1892661    19461    66511     4.0  1284593250
1973526    20179    66511     2.0  1345422584
2225496    22870    66511     3.0  1359449954
2237277    22964    66511     5.0  1294864768
2697982    27860    66511     3.0  1535759538
...          ...      ...     ...         ...
26813746  273795    66511     3.0  1375648973
26871575  274367    66511     3.5  1299940736
27121778  276793    66511     3.0  1306264404
27252068  278103    66511     3.5  1477342170
27616760  281790    66511     3.0  1322864758

[89 rows x 4 columns]

> There are 89 entries found for the same movieId but it have other column values different so these are not duplicated

Check entries for other movieId of the same title "Berlin Calling (2008)" id=191775

In [26]:
ratings[ratings.movieId == 191775]

userId  movieId  rating   timestamp
4190621    43044   191775     3.5  1534967484
15654247  159846   191775     2.0  1534046590
25704535  262584   191775     3.5  1532697651

> There are 3 entries which are associated with this movieId which is for same movie that have 89 entries but with different movieId.

After this investigation, write a function to replace movieId in rating dataframe

In [27]:
def replace_movie_id_in_rating_dataframe(dataframe: pd.DataFrame, new_movie_id: int, old_movie_id: int) \
    -> pd.DataFrame:
    dataframe.loc[dataframe.movieId == old_movie_id, 'movieId'] = new_movie_id
    return dataframe

Let's try this function on same movie "Berlin Calling (2008)" which have two movieIds

In [28]:
ratings = replace_movie_id_in_rating_dataframe(ratings, 66511, 191775)

Now Check the entries in ratings dataframe, whether it have entry for old movie Id (Here we replace 191775 with 66511)

In [29]:
(ratings.movieId == 191775).sum()

0

> All entries associated with old entry has been replaced with new one as count is zero.

Verify the count associated with new movieId

In [30]:
(ratings.movieId == 66511).sum()

92

> We can see that now the old entries are merged into new id for which count was 89 before and now it's 92.

Voila! Our function is working!

Now We will write a function to remove the old Ids that we replaced, now from movies dataset.

In [31]:
def remove_old_movie_ids_from_movies_dataframe(dataframe: pd.DataFrame, old_movie_ids: [int]) -> pd.DataFrame:
    target_index = dataframe[dataframe.movieId.isin(old_movie_ids)].index
    dataframe = dataframe.drop(target_index)
    return dataframe

Let's try this function on sample values on the dataset.

In [32]:
remove_old_movie_ids_from_movies_dataframe(movies, [1, 2, 3, 4])

movieId                                    title  \
4            5       Father of the Bride Part II (1995)   
5            6                              Heat (1995)   
6            7                           Sabrina (1995)   
7            8                      Tom and Huck (1995)   
8            9                      Sudden Death (1995)   
...        ...                                      ...   
58093   193876                  The Great Glinka (1946)   
58094   193878  Les tribulations d'une caissière (2011)   
58095   193880                 Her Name Was Mumu (2016)   
58096   193882                             Flora (2017)   
58097   193886                              Leal (2018)   

                              genres  
4                             Comedy  
5              Action|Crime|Thriller  
6                     Comedy|Romance  
7                 Adventure|Children  
8                             Action  
...                              ...  
58093             (no genres listed)  
58094                         Comedy  
58095                          Drama  
58096  Adventure|Drama|Horror|Sci-Fi  
58097             Action|Crime|Drama  

[58094 rows x 3 columns]

> As we can see now the index is starting from 4 as first 4 movieIds are removed from the dataset. As this was only a sample execution, we will not update our original dataset with this updated values.

As now we are sure about the execution of the function, Let's make it execute for all duplicated movie Ids, where we will keep first appeared movieId as original one which we will keep and other all instances with same movie title will be replaced and collect all duplicated ids in a list defined which will be appended after each iteration which will be input for second function to remove same ids from movies dataset.

In [33]:
def removal_of_ids_from_both_dataframes(movies_dataframe: pd.DataFrame, ratings_dataframe: pd.DataFrame) -> (pd.DataFrame, pd.DataFrame):
    duplicates = movies_dataframe[movies_dataframe.duplicated(subset=['title', 'genres'], keep=False)]
    old_movie_ids = []
    for movie_title in duplicates.title.unique():
        movie_ids = duplicates[duplicates.title == movie_title].movieId.values
        new_movie_id, old_movie_id = movie_ids
        ratings_dataframe = replace_movie_id_in_rating_dataframe(ratings_dataframe, new_movie_id, old_movie_id)
        old_movie_ids.append(old_movie_id)
    movies_dataframe = remove_old_movie_ids_from_movies_dataframe(movies_dataframe, old_movie_ids)
    return movies_dataframe, ratings_dataframe

Check actual execution on datasets and save the results in movies_df and ratings_df respectively

In [34]:
movies_df, ratings_df = removal_of_ids_from_both_dataframes(movies, ratings)

After successful execution, Let's check some details to confirm whether removal and updating datasets worked or not

In [35]:
len(movies_raw) - len(movies_df)

14

> Voila! It's correct. movies original dataset has 58098 total entries among which when calculated, 14 were duplicated in terms of column. After execution of this function the duplicated entries should be removed. So the difference between original and updated dataframe should be 14 and it is the same!

Check the same for ratings dataset

In [36]:
len(ratings_raw) - len(ratings_df)

0

> For ratings dataset, there were different movieIds were noted for the same movie but it was not a duplication. So there should not be any removal in ratings dataset.
Here we did replace the movieIds with only one instance for each movie so that the difference between original and updated dataset is 0.

## Operations on Ratings

In [45]:
ratings_df

userId  movieId  rating   timestamp
0              1      307     3.5  1256677221
1              1      481     3.5  1256677456
2              1     1091     1.5  1256677471
3              1     1257     4.5  1256677460
4              1     1449     4.5  1256677264
...          ...      ...     ...         ...
27753439  283228     8542     4.5  1379882795
27753440  283228     8712     4.5  1379882751
27753441  283228    34405     4.5  1379882889
27753442  283228    44761     4.5  1354159524
27753443  283228    54286     4.5  1354159718

[27753444 rows x 4 columns]

Now let's re-ensure duplication in dataset if there any

In [46]:
ratings_df.duplicated(keep=False).sum()

0

Checking duplication with respect to columns

In [47]:
ratings_df.duplicated(subset=['userId','movieId', 'rating', 'timestamp']).sum()

0

> So there are no duplicates present in new dataset we have

Let's try by removing timestamp column as it is not important when we consider user and movies and their ratings.
Create a new dataset by selecting only 'userId', 'movieId', 'rating' from ratings_df.

In [43]:
ratings_df_1 = ratings_df[['userId', 'movieId', 'rating']]
ratings_df_1

userId  movieId  rating
0              1      307     3.5
1              1      481     3.5
2              1     1091     1.5
3              1     1257     4.5
4              1     1449     4.5
...          ...      ...     ...
27753439  283228     8542     4.5
27753440  283228     8712     4.5
27753441  283228    34405     4.5
27753442  283228    44761     4.5
27753443  283228    54286     4.5

[27753444 rows x 3 columns]

Further let's try to get count of total number of duplicates

In [44]:
ratings_df_1.duplicated(subset=['userId','movieId', 'rating']).sum()

13

> Without including timestamp now, There are 13 duplicates found

Check the duplicated entries and verify

In [48]:
ratings_df_2 = ratings_df_1[ratings_df_1.duplicated(subset=['userId','movieId', 'rating'], keep=False)]
ratings_df_2

userId  movieId  rating
6213836    63786   104155     0.5
6214019    63786   104155     0.5
7128279    73145   104155     4.0
7128522    73145   104155     4.0
10394207  106859    70155     3.0
10394847  106859    70155     3.0
10862664  111688   104155     4.5
10862692  111688   104155     4.5
12010931  123100    97773     3.0
12011347  123100    97773     3.0
12243097  125365   111519     4.5
12243194  125365   111519     4.5
15264429  155976    48682     3.5
15264776  155976    48682     3.5
15653605  159846    66511     2.0
15654247  159846    66511     2.0
16408991  167835   104155     3.0
16409115  167835   104155     3.0
19174782  195628   104155     3.0
19174990  195628   104155     3.0
19242274  196323     2644     5.0
19243096  196323     2644     5.0
22505207  230183    97773     3.5
22505255  230183    97773     3.5
26353900  269098   104155     3.5
26354146  269098   104155     3.5

> So we have above entries with same row values which we can reduce and only keep single entry from that.

Now we will try to operate on removal of this duplication for these 13 row values

In [49]:
ratings_df_2.userId.unique()

array([ 63786,  73145, 106859, 111688, 123100, 125365, 155976, 159846,
       167835, 195628, 196323, 230183, 269098], dtype=int64)

Before going to global change we will first test our logic on single value entry

In [52]:
ratings_df_2[ratings_df_2.userId == 63786]

userId  movieId  rating
6213836   63786   104155     0.5
6214019   63786   104155     0.5

> There are two entries present among which we will keep the first row and remove the second row

Try to get an index value for the row 

In [60]:
ratings_df_2[ratings_df_2.userId == 63786].index.values.astype(int)[1]

6214019

Now using drop in pandas, Drop the row with index 6214019

In [62]:
ratings_df_2[ratings_df_2.userId == 63786].drop(6214019)

userId  movieId  rating
6213836   63786   104155     0.5

We dropped 6214019 index row values successfully so let's try to make function to do the same globally

In [68]:
# Just make sure its working for ratings_df_1 containing all entries
test = ratings_df_1.drop(6214019)

In [69]:
# Check whether the index is deleted or not
test[test.index.values.astype(int) == 6214019]

Empty DataFrame
Columns: [userId, movieId, rating]
Index: []

Super! Our test is working, Let's build a function for global removal of these entries on our dataset.

In [85]:
def remove_duplicate_row_values_from_ratings_dataset(dataframe: pd.DataFrame, user_ids: [int]) -> pd.DataFrame:
    for user_id in user_ids:        
        index_to_remove = dataframe[dataframe.userId == user_id].index.values.astype(int)[1]
        dataframe = dataframe.drop(index_to_remove)
    return dataframe

Execute the function and save it to variable named ratings_df_final

In [91]:
ratings_df_final = remove_duplicate_row_values_from_ratings_dataset(ratings_df_1, ratings_df_2.userId.unique())

It Works! Let's make sure the deletion of duplicated entries by comparing size of previous dataset with new updated dataset

In [92]:
len(ratings_df_1) - len(ratings_df_final)

13

> Voila! Our function is working as the difference between size of old and new dataset is same as the number of duplicate entries were present. 

We have successfully removed duplicate rows from our dataset.
Now we can process our analysis further on "movies_df" and "ratings_df_final"